### PR TITLE
SEA3D and AnimationClip update 

### PR DESCRIPTION
**Pull request by sunag**
- Optimization: Unify SEA3D `AnimationMixer`
- `THREE.AnimationMixer` dictionary based in `AnimationClip.uuid` preventing name conflicts
- Fix cross-browser data.slice to data.subarray
- New features in `webgl_loader_sea3d_skinning` complex clone with middle click ( compatible with sea3d script ).
- Fix deprecated `warns` in set `shadow` and `sound` loader

The `AnimationMixer` update is compatible with the previous examples. 
The recommendation is always use:

``` javascript
mixer.clipAction( animationClip ).. .play()
```

and not:

``` javascript
mixer.clipAction( "animationName" ).. .play()
```

Mainly if the `mixer` have many `animation clips`.

### CODE REVIEW

**Comment by mrdoob** [comment]
Oh. Seems like there are some conflicts. Probably with the sound example. Do you mind resolving these in the PR?

@tschw do the animation changes look good?

**Comment by tschw** [comment]
Reading...

File: src/animation/AnimationMixer.js [hunk 1]
```diff
@@ -83,8 +83,10 @@ Object.assign( THREE.AnimationMixer.prototype, THREE.EventDispatcher.prototype,
 
 		var root = optionalRoot || this._root,
 			rootUuid = root.uuid,
-			clipName = ( typeof clip === 'string' ) ? clip : clip.name,
-			actionsForClip = this._actionsByClip[ clipName ];
+			clipObject = typeof clip === 'string' ? this.getClipByName( clip ) : clip,
```
Comment (tschw) [inline src/animation/AnimationMixer.js]: @sunag 
There's this issue that multiple objects can have different clips sharing the same name and it's actually not all too uncommon: 
Different characters that can all perform the same actions may e.g. "walk" differently an thus use different clips with that same name.

Instead of searching through existing actions, I guess it would be best to do something like `AnimationClip.findByName( root.animationClips, clip )`. Note that there's `AnimationClip.findByName`, but I'm not sure whether `root.animationClips` exists, or what its name is in case it does, or whether it would be the only possible name :). However, the nice thing about this approach would be that you could always pass a name to the mixer and it will always give you an action, currently playing or not, and create an action on demand, even if the clip has never been passed-in directly. I guess this may simplify higher level code quite a bit.

A less preferable fix would be to make it `this._getClipByNameAndRoot( clip, rootUuid )` and also compare against `action.getRoot().uuid` searching through the actions. This way the collision is resolved, but the limitation that the mixer must first see a clip "by object" before it can be addressed by name would remain. Also, the linear search through all cached actions is somewhat undesirable (there can be many of them).

Comment (bhouston) [inline src/animation/AnimationMixer.js]: The way Mixer was intended to be used is to have one per character.  Thus you can tell a character to walk and it would run the appropriate clip for that character as it would be preloaded into its Mixer under that name.

We have sort of diverged from that intended design to a global Mixer, which really wasn't intended.  I allowed for a Mixer to have multiple roots so you could animate something in a characters hand -- it would be a separate root.  Not really so that you could have a centralized Mixer that controls multiple characters.

Unity and UE4 have one mixer per character.  And they even have more complex structures built on the per-character mixers to make more complex things.  I am little concerned that if we continue down this global mixer path we'll continue to run into design issues like this one as this isn't how things are usually done.

Comment (bhouston) [inline src/animation/AnimationMixer.js]: Basically we are misusing the Mixer as a global Mixer when it was intended to be a per-character mixer with support for extra synchronized objects with that character.

Softimage has a per-character mixer that supports clips: http://softimage.wiki.softimage.com/xsidocs/animmixer1.htm

Unity3D has a per-character mixer that supports clips: http://docs.unity3d.com/ScriptReference/Animation.html

Maya has a per character mixer: http://help.autodesk.com/view/MAYAUL/2016/ENU/?guid=GUID-B90726CA-6C86-4E9B-80F3-1F65745C3B3F

UE4 has a per-character animation Blueprint and a per-character AnimGraph: https://docs.unrealengine.com/latest/INT/Engine/Animation/AnimBlueprints/

Basically we are being non-standard in our usage of a Global Mixer and it is going to lead to continued issues because it is not the appropriate abstraction to the problem.

Comment (bhouston) [inline src/animation/AnimationMixer.js]: I m okay if we continue down this road, but be aware that we will need to eventually create a new per-character mixer class that makes using this global mixer easier for individual characters.  It will allow one to pre-load animations into it by name and play them back by name.  I think it can be build onto of the global mixer relatively easily -- just it wasn't my intention to have a global mixer at all, but I guess we can work with it.

We always get to the right designs eventually, although sometimes we take circuitous routes to them, and if this is how we are going to get there, I am okay with that, but just be aware that we have to get back to where there are per-character mixers of some sort.

**Comment by sunag** [comment]
@tschw Am considering use `AnimationMixer` per character so I can we can think better in a future solution for this. Maybe a possible `AnimationGlobalMixer` or `AnimationMixers`!? 

There also the possibility of add more one dictionary for the names thus not need searching of clip in existing actions increasing the performance. `AnimationMixer` would both to `local` and `global` animations.

File: src/animation/AnimationMixer.js (hunk 1 above)
Comment (tschw) [inline src/animation/AnimationMixer.js]: @bhouston 
As long as all characters are different there shouldn't be a problem using multiple mixers. 

When you have many instances of a single character / object (e.g. mobs in an FPS game) we'd currently need one mixer to handle all of them for reasonable resource consumption.

The potential name conflicts are worth resolving in any case, I think.

Comment (tschw) [inline src/animation/AnimationMixer.js]: @bhouston 
> I am okay if we continue down this road, but be aware that we will need to eventually create a new per-character mixer class that makes using this global mixer easier for individual characters.

This might be something @sunag is actually building.

Considering the current functionality, and that most of it sits on the actions anyway, a per-character mixer would be a trivial wrapper binding a default for `localRoot`. Am I missing something?

Comment (bhouston) [inline src/animation/AnimationMixer.js]: > Considering the current functionality, and that most of it sits on the actions anyway, a per-character mixer would be a trivial wrapper binding a default for localRoot. Am I missing something?

I think that is it.  Well, except that one loads animation clips into the local-per character mixers and then play them with the simple names like "walk".  Playing the "walk" animation would resolve to different animation clips based on which per-character mixer you play it from -- because it resolves to the local per-character walk clip, a global clip named walk.  Thus it maintains the localRoot and maintains the resolving of the local animation names to their animation clips, which is usually done at load time by initializing the per-character mixer with the animation clips and their local names.  I think if we did that, such a wrapper would be equivalent to the basic per-character mixers idea.

**Comment by tschw** [comment]
@sunag 
As said, when a name is specified, the mixer should ideally look up the clip on the root object and we should define a place for clips to live.

They actually belong onto the geometry.

Unfortunately the data model is properly flawed, since it puts them onto each object: When you save a scene to JSON with three instances of the same character, you triple all animation data. Conversely, for loading it has to be present three times and each redundant clip will get its own UUID :(.

But since we're talking SEA3D, I think it shouldn't keep us from doing the right thing. 
1. First look for `root.geometry.animations`
2. Look for `root.animations` (this can be removed when the three.js data model is fixed)
3. Search for a (clip,root) tuple in the actions (this can be removed when the three.js data model is fixed).

I can't quite parse your first sentence.

I'd go with the global mixer for now and always pass the object as the second argument to `clipAction`. When we implement a local mixer we simply change the constructor and remove that argument - that's it.

**Comment by sunag** [comment]
> A less preferable fix would be to make it this._getClipByNameAndRoot( clip, rootUuid ) and also compare against action.getRoot().uuid searching through the actions. 

Updated!

File: src/animation/AnimationMixer.js (hunk 1 above)
Comment (tschw) [inline src/animation/AnimationMixer.js]: > Am I missing something?

Yep, there's also `stopAllAction` (which is clearly very important) and `update`.

@bhouston 
I guess we can go without the latter and simply have the global time update be global?

**Comment by bhouston** [comment]
@tschw wrote:
> Unfortunately the data model is properly flawed, since it puts them onto each object: When you save a scene to JSON with three instances of the same character, you triple all animation data. Conversely, for loading it has to be present three times and each redundant clip will get its own UUID :(.

Actually I supported both the idea of scene animations and character animations.  You can have animations on the scene, for things like cinematics:

https://github.com/mrdoob/three.js/blob/master/src/loaders/ObjectLoader.js#L64

You can also have animations on geometry as well that can be shared among instances of that geometry:

https://github.com/mrdoob/three.js/blob/master/src/loaders/JSONLoader.js#L99
https://github.com/mrdoob/three.js/blob/master/src/loaders/JSONLoader.js#L537

I supported both cases because this is generally want you want in a real-world tool -- scene animations and character animations.

**Comment by bhouston** [comment]
To be clear, I only introduced the new animation format where it didn't exist before, on the scene for cinematics, while keeping full backwards compatibility with the old animation format on the geometry so that we didn't have to update our exporters to take advantage of the new mixer capability.  Yes, it lead to inconsistencies in the format, but incremental updates while maintaining backwards compatibility seemed to be the easiest way forward.

**Comment by tschw** [comment]
@bhouston 
> To be clear, I only introduced the new animation format where it didn't exist before, on the scene for cinematics, while keeping full backwards compatibility with the old animation format on the geometry so that we didn't have to update our exporters to take advantage of the new mixer capability.

Ah! OK. Turns out it's not as flawed as I guesstimated from quickly (mis)reading the code. Since it's only allowed on the root, there's no duplication going on. Unfortunately the other loader states it's the "old" way of sorts, which is what got me confused.

So the correct way to find a clip on an object is `root.animations` when it's a `Scene` and `root.geometry.animations` otherwise?

**Comment by bhouston** [comment]
> So the correct way to find a clip on an object is root.animations when it's a Scene and root.geometry.animations otherwise?

I would check both locations because then you can put animations on any object or any geometry and it would work.  I think that one may want to put animations on mesh nodes or other nulls that are not the scene.

**Comment by sunag** [comment]
> Oh. Seems like there are some conflicts. Probably with the sound example. Do you mind resolving these in the PR?

I'm not sure how to solve this. Maybe it's easier I create a new PR?

**Comment by tschw** [comment]
I'm currently making a PR for it. I pulled f1e9e16 off this branch.

**Comment by sunag** [comment]
@mrdoob @tschw @bhouston I will close the PR and open a new just with the SEA3D updates.

**Comment by tschw** [comment]
@sunag 
> I'm not sure how to solve this. Maybe it's easier I create a new PR?

```shell
$ git checkout dev
$ git branch feature
$ git log # see where the branch starts
$ git reset --hard <ref before branch>
$ git pull upstream dev
$ git checkout feature
$ git rebase dev
$ # now everytime you want to push your changes:
$ git checkout dev
$ git reset --hard feature
$ git push origin --force dev
$ git log # see where the rewritten branch starts
$ git reset --hard <ref before feature branch>
$ git checkout feature
```

That's why we don't recommend using your `dev` branch for PRs :).

In order to remove individual commits you can use `git rebase -i <ref before branch>` then an editor opens with all the commits in your branch and you can simply delete the commit and your branch will be rewritten without it.

**Comment by sunag** [comment]
> That's why we don't recommend using your dev branch for PRs :).

Thank you by all explanation. I'll do it from now on : )
